include/sys/icu.h

```diff
@@ -0,0 +1,36 @@
+#ifndef _ICU_METHODS_H_
+#define _ICU_METHODS_H_
+
+#include <sys/device.h>
+#include <sys/resource.h>
+#include <sys/interrupt.h>
+
+typedef struct icu_methods {
+  void (*intr_setup)(device_t *dev, resource_t *irq, ih_filter_t *filter,
+                     ih_service_t *service, void *arg, const char *name);
+  void (*intr_teardown)(device_t *dev, resource_t *irq);
+} icu_methods_t;
+
+#define ICU_METHODS(dev) (*(icu_methods_t *)(dev)->driver->interfaces[DIF_ICU])
+
+static inline void icu_intr_setup(device_t *dev, resource_t *irq,
+                                  ih_filter_t *filter, ih_service_t *service,
+                                  void *arg, const char *name) {
+  assert(dev->parent);
+  if (!dev->parent->driver->interfaces[DIF_ICU]) {
+    icu_intr_setup(dev, irq, filter, service, arg, name);
```

Comment:
It's broken and leads to an infinite loop. You probably want to call `icu_intr_setup(dev->parent, irq, filter, service, arg, name)`, but it's not that simple. For now, rootdev and GT64120 assume that a device which requests an interrupt setup is a direct child of the bus, eg:
```C
gt_pci_intr_setup(...) {
  assert(dev->parent->driver == &gt_pci_bus);
  ...
}
rootdev_intr_setup(...) {
  rootdev_t *rd = dev->parent->state;
  ...
}
```

Reply:
You are right! I must've accidentally change it from `dev->parent` to `dev` when I was willing to change something else. I'll correct it soon.

Reply:
Fixed.